local_planner/src/nodes/planner_functions.cpp

```diff
@@ -355,74 +355,25 @@ void costFunction(float e_angle, float z_angle, float obstacle_distance, const E
-    path_node_positions_extended.insert(path_node_positions_extended.end(), path_node_positions.begin(),
-                                        path_node_positions.end());
-    int size_extended = path_node_positions_extended.size();
-
-    // find path nodes between which the drone is currently located:
-    // a vector with the distances of each node to the drone is generated
-    // the indices of the nodes with smallest and next smallest distance are
-    // found
-    int min_dist_idx = 0;
-    int second_min_dist_idx = 0;
-    float min_dist = HUGE_VAL;
-    float second_min_dist = HUGE_VAL;
-    std::vector<float> distances;
-    distances.reserve(size_extended);
-
-    for (int i = 0; i < size_extended; i++) {
-      distances.push_back((position - path_node_positions_extended[i]).norm());
-      if (distances[i] < min_dist) {
-        second_min_dist_idx = min_dist_idx;
-        second_min_dist = min_dist;
-        min_dist = distances[i];
-        min_dist_idx = i;
-      } else if (distances[i] < second_min_dist) {
-        second_min_dist = distances[i];
-        second_min_dist_idx = i;
-      }
-    }
+bool getSetpointFromTree(const std::vector<Eigen::Vector3f>& tree, const ros::Time& tree_generation_time,
+                         float velocity, Eigen::Vector3f& setpoint) {
+  int i = tree.size();
+  // path contains nothing meaningful
+  if (i < 2) {
+    return false;
+  }
 
-    // the drone is located between the two nodes(min_dist_idx,
-    // second_min_dist_idx),
-    // wp_idx describes the node further ahead in the path
-    int wp_idx = std::min(min_dist_idx, second_min_dist_idx);
-
-    // if drone is too far away from tree or already at last node -> don't use
-    // tree
-    if (min_dist > 3.0f || wp_idx == 0) {
-      tree_available = false;
-    } else {
-      float cos_alpha = (node_distance * node_distance + distances[wp_idx] * distances[wp_idx] -
-                         distances[wp_idx + 1] * distances[wp_idx + 1]) /
-                        (2.0f * node_distance * distances[wp_idx]);
-      float l_front = distances[wp_idx] * cos_alpha;
-      float l_frac = l_front / node_distance;
-
-      Eigen::Vector3f mean_point =
-          (1.f - l_frac) * path_node_positions_extended[wp_idx - 1] + l_frac * path_node_positions_extended[wp_idx];
-
-      p_pol = cartesianToPolarHistogram(mean_point, position);
-      p_pol.r = 0.0f;
-    }
-  } else {
-    tree_available = false;
+  // step through the tree until the point where we should be if we had traveled perfectly with velocity along it
+  float distance_left = std::max(0.1, (ros::Time::now() - tree_generation_time).toSec() * velocity);
+  setpoint = tree[i - 2];
+  Eigen::Vector3f tree_segment = tree[i - 2] - tree[i - 1];
+
+  for (i = tree.size() - 2; i > 0 && distance_left > tree_segment.norm(); --i) {
+    distance_left -= tree_segment.norm();
+    tree_segment = tree[i - 1] - tree[i];
+    setpoint = tree[i - 1];
```

Comment:
I think we should do a linear interpolation for the final points, otherwise we get sharp jumps in the setpoint

Reply:
now we're sending a setpoint one segment ahead, which should remove those jumps